### PR TITLE
Support LLVM 19.1

### LINKED ISSUE #14840: LLVM 19 support
[LLVM 19.1.0-rc1 has been released](https://github.com/llvm/llvm-project/releases/tag/llvmorg-19.1.0-rc1). There are some breaking changes there:

* [3DNow! is dropped](https://github.com/llvm/llvm-project/blob/release/19.x/llvm/docs/ReleaseNotes.rst#changes-to-the-x86-backend). Crystal has never supported it anyway, but this might affect #14524 depending on how we gather the LLVM subtarget info
* [`LLVMDIBuilderInsertDeclareAtEnd` is deleted](https://github.com/llvm/llvm-project/blob/release/19.x/llvm/docs/ReleaseNotes.rst#changes-to-the-c-api), the replacement is `LLVMDIBuilderInsertDeclareRecordAtEnd`
* [The MMX register type is dropped on the 20.x development branch](https://github.com/llvm/llvm-project/blob/main/llvm/docs/ReleaseNotes.rst#changes-to-the-c-api), this probably affects the `LLVM::Type::Kind` enum

### PR DESCRIPTION
**Pull request by HertzDevil**
Closes #14840. Tested on Debian using `1:19.1.0~++20240730073008+404746b9f21b-1~exp1~20240730073118.4`.

The return type of `LLVM::DIBuilder#insert_declare_at_end` now differs between `LibLLVM::ValueRef` and `LibLLVM::DbgRecordRef` depending on whether LLVM is new enough (both are typedefs to `Void*`), although this return value is never used in the compiler itself.

### CODE REVIEW

**Review by straight-shoota — APPROVED** [review]

File: src/llvm/di_builder.cr [hunk 1]
```diff
@@ -95,7 +95,11 @@ struct LLVM::DIBuilder
   end
 
   def insert_declare_at_end(storage, var_info, expr, dl : LibLLVM::MetadataRef, block)
-    LibLLVM.di_builder_insert_declare_at_end(self, storage, var_info, expr, dl, block)
+    {% if LibLLVM::IS_LT_190 %}
+      LibLLVM.di_builder_insert_declare_at_end(self, storage, var_info, expr, dl, block)
+    {% else %}
+      LibLLVM.di_builder_insert_declare_record_at_end(self, storage, var_info, expr, dl, block)
+    {% end %}
```
Comment (ysbaddaden) [inline src/llvm/di_builder.cr:102]: Oh my, we're returning a raw `Void*` from LLVM here, which may be a `LibLLVM::ValueRef` or a `LibLLVM::DbgRecordRef` now :disappointed:

**Comment by HertzDevil** [comment]
[19.1.0 stable has been released](https://github.com/llvm/llvm-project/releases/tag/llvmorg-19.1.0); awaiting the volunteer packages for CI

**Comment by HertzDevil** [comment]
It appears there are now official binaries by release managers and all the third-party ones are only on their forum: https://discourse.llvm.org/t/rfc-improve-binary-security/78121/56

File: .github/workflows/llvm.yml [hunk 2]
```diff
@@ -17,17 +17,19 @@ jobs:
...
+            llvm_filename: "clang+llvm-15.0.6-x86_64-linux-gnu-ubuntu-18.04.tar.xz"
           - llvm_version: "16.0.3"
-            llvm_ubuntu_version: "22.04"
+            llvm_filename: "clang+llvm-16.0.3-x86_64-linux-gnu-ubuntu-22.04.tar.xz"
           - llvm_version: "17.0.6"
-            llvm_ubuntu_version: "22.04"
+            llvm_filename: "clang+llvm-17.0.6-x86_64-linux-gnu-ubuntu-22.04.tar.xz"
           - llvm_version: "18.1.4"
-            llvm_ubuntu_version: "18.04"
+            llvm_filename: "clang+llvm-18.1.4-x86_64-linux-gnu-ubuntu-18.04.tar.xz"
```
Comment (straight-shoota) [inline .github/workflows/llvm.yml:30]: **praise:** I like this change. It's a bit more boilerplate, but also more expressive than the previous version. I would probably prefer this even if it wasn't necessitated by a change in the naming scheme.

File: src/compiler/crystal/codegen/debug.cr [hunk 3]
```diff
@@ -367,6 +367,16 @@ module Crystal
...
+        # FIXME: When debug records are used instead of debug intrinsics, it
+        # seems inserting them into an empty BasicBlock will instead place them
+        # in a totally different (next?) function where the variable doesn't
+        # exist, leading to a "function-local metadata used in wrong function"
+        # validation error. This might happen when e.g. all variables inside a
+        # block are closured. Ideally every debug record should immediately
+        # follow the variable it declares.
+        {% unless LibLLVM::IS_LT_190 %}
+          call(do_nothing_fun) if block.instructions.empty?
+        {% end %}
```
Comment (straight-shoota) [inline src/compiler/crystal/codegen/debug.cr:379]: **thought:** This really seems quite a bit strange. But if it works... 🤷

File: src/llvm/di_builder.cr (hunk 1 above)
Comment (straight-shoota) [reply src/llvm/di_builder.cr:103]: There's only a single call to this method in the compiler, and the return value is ignored. Maybe we should tack on a type restriction to `Nil`? :shrug:

**Review by straight-shoota — APPROVED** [review]

File: src/llvm/di_builder.cr (hunk 1 above)
Comment (ysbaddaden) [reply src/llvm/di_builder.cr:103]: For the crystal compiler we may not care, but the method is public...

Comment (HertzDevil) [reply src/llvm/di_builder.cr:103]: Or we do the sensible thing of wrapping those pointers into an actual `LLVM::Metadata` or `LLVM::DebugRecord` struct

Comment (HertzDevil) [reply src/llvm/di_builder.cr:103]: Also `LLVM::DIBuilder` is marked as `@[Experimental]`, so we are free to make breaking changes here, and probably not in a hurry to do so

**Comment by straight-shoota** [comment]
I think it's fine to still add this into 1.14.0 even though we're already in the freeze period. The changes only have an effect when linking against LLVM 1.19, so this should not affect any existing configuration.

**Pull request merged**